src/openclsim/critical_path/simulation_graph.py

```diff
@@ -64,36 +150,286 @@ def __check_dep(self, dependency_list):
         """
-        return nx.DiGraph()
+        self.simulation_graph = nx.DiGraph()
 
-    def __find_critical_edges(self):
+        self.__create_activity_edges()
+        self.__link_activity_edges()
+
+        # check if the resulting graph is acyclic
+        assert nx.is_directed_acyclic_graph(
+            self.simulation_graph
+        ), "The resulting graph appears to be cyclic!"
+
+        return self.simulation_graph
+
+    def __create_activity_edges(self):
         """
-        Iteratively run through self.simulation_graph and find all critical edges
+        Create nodes and edges for each individual activity.
+
+        Converts all cp_activity_ids within ``self.recorded_activity_df`` into start and end nodes
+        connected through an edge. All relevant attributes, such as the
+        duration of an activity, are added to these nodes and edges.
         """
-        pass
+        cp_activities_df = self.recorded_activity_df.drop_duplicates(
+            subset=["cp_activity_id"]
+        )
+        for params in cp_activities_df.itertuples():
+            # names of the nodes by start/end of
+            name_start = f"{self.__NODE_START_PREFIX} {params.cp_activity_id}"
+            name_end = f"{self.__NODE_END_PREFIX} {params.cp_activity_id}"
+
+            # add the start node
+            self.simulation_graph.add_node(
+                name_start,
+                pos=(params.start_time, params.Index),
+                time=params.start_time,
+                cp_activity_id=params.cp_activity_id,
+            )
+
+            # add the end node
+            self.simulation_graph.add_node(
+                name_end,
+                pos=(params.end_time, params.Index),
+                time=params.end_time,
+                cp_activity_id=params.cp_activity_id,
+            )
+
+            # add the edge
+            kwargs = {
+                "node_start": name_start,
+                "node_end": name_end,
+                "activity": params.activity,
+                "start_time": params.start_time,
+                "state": params.state,
+                "duration": params.duration,
+                "end_time": params.end_time,
+                "cp_activity_id": params.cp_activity_id,
+                "edge_type": self.__EDGE_TYPES[0],
+                self.__COLNAME_CRITICAL: self.__CRITICAL[0],
+            }
+            self.simulation_graph.add_edge(name_start, name_end, **kwargs)
+
+    def __link_activity_edges(self):
+        """
+        Add connecting edges between linked activities.
+
+        Note: this method is to be called after
+        ``self.__create_activity_edges()``.
+
+        Uses the ``self.list_dependencies`` to link all loose nodes and edges
+        together within the graph. Note that edges are added from the END of
+        the first activity to the START of the second.
+        """
+        for name_start, name_end in self.dependency_list:
```

Comment:
what is the meaning of `name_start` and `name_end` here?

Reply:
Ah I see, the name of the start and the end node to be connected. I was a bit confused, since `name_start` is actually the name of the *end* node of the first activity, and `name_end` is actually the name of the *start* node of the second activity...

Maybe we should change this so that there is no ambiguity here in the code